File: lib-segment/src/main/java/websiteschema/mpsegment/core/POSResources.java
package websiteschema.mpsegment.core;

import websiteschema.mpsegment.conf.MPSegmentConfiguration;
import websiteschema.mpsegment.hmm.NodeRepository;
import websiteschema.mpsegment.hmm.Pi;
import websiteschema.mpsegment.hmm.Transition;
import websiteschema.mpsegment.util.FileUtil;
import websiteschema.mpsegment.util.SerializeHandler;

import java.io.DataInputStream;
import java.io.IOException;

/**
 * Created with IntelliJ IDEA.
 * User: twer
 * Date: 9/17/12
 * Time: 8:31 PM
 * To change this template use File | Settings | File Templates.
 */
public class POSResources {

    private static POSResources instance = new POSResources();

    private int[] posFreq;
    private Transition transition;
    private Pi pi;
    private NodeRepository stateBank;

    public static POSResources getInstance() {
        return instance;
    }

    POSResources() {
        initialize();
    }

    private void initialize() {
        transition = new Transition();
        pi = new Pi();
        stateBank = new NodeRepository();

        String resource = MPSegmentConfiguration.getINSTANCE().getPOSMatrix();
        try {
            SerializeHandler readHandler = new SerializeHandler(
                    new DataInputStream(FileUtil.getResourceAsStream(resource)));
            load(readHandler);
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    private void load(SerializeHandler readHandler) throws IOException {
        posFreq = readHandler.deserializeArrayInt();
        transition.load(readHandler);
        pi.load(readHandler);
        stateBank.load(readHandler);
    }

    public int[] getPosFreq() {
        return posFreq;
    }

    public Transition getTransition() {
        return transition;
    }

    public Pi getPi() {
        return pi;
    }

    public NodeRepository getStateBank() {
        return stateBank;
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/core/POSResources.scala
package websiteschema.mpsegment.core

import websiteschema.mpsegment.conf.MPSegmentConfiguration
import websiteschema.mpsegment.hmm._
import websiteschema.mpsegment.util.FileUtil
import websiteschema.mpsegment.util.SerializeHandler

import java.io.DataInputStream

object POSResources {

  private var posFreq: Array[Int] = null
  private var transition: Transition = null
  private var pi: Pi = null
  private val root = new Trie()
  private var stateBank: NodeRepository = null

  initialize()

  private def initialize() {
    transition = Transition()
    pi = Pi()
    stateBank = new NodeRepository()

    val resource = MPSegmentConfiguration().getPOSMatrix()
    try {
      val readHandler = SerializeHandler(new DataInputStream(FileUtil.getResourceAsStream(resource)))
      load(readHandler)
    } catch {
      case ex: Throwable =>
        ex.printStackTrace()
    }
  }

  private def load(readHandler: SerializeHandler) {
    posFreq = readHandler.deserializeArrayInt()
    loadTransition(readHandler)
    pi.load(readHandler)
  }

  private def loadTransition(readHandler: SerializeHandler) {
    root.load(readHandler)
    stateBank.load(readHandler)
  }

  def save(writeHandler: SerializeHandler) {
    writeHandler.serializeArrayInt(posFreq)
    root.save(writeHandler)
    stateBank.save(writeHandler)
    pi.save(writeHandler)
  }

  def getPosFreq(): Array[Int] = {
    return posFreq
  }

  def getTransition(): ITransition = {
    val tran = new BigramTransition(50)
    root.descendant.foreach(node => {
      node.descendant.foreach(child => {
        if (0 < node.getKey() && 0 < child.getKey()) {
          tran.setProb(node.getKey(), child.getKey(), child.getProb())
        }
      })
    })

    return tran
  }

  def getPi(): Pi = {
    return pi
  }

  def getStateBank(): NodeRepository = {
    return stateBank
  }
}
